src/Symfony/Component/DependencyInjection/Dumper/Preloader.php

```diff
@@ -80,21 +95,104 @@ private static function doPreload(string $class, array &$preloaded)
                         $c = $p->getDefaultValueConstantName();
 
                         if ($i = strpos($c, '::')) {
-                            self::doPreload(substr($c, 0, $i), $preloaded);
+                            self::doPreload(substr($c, 0, $i), $preloaded, $loader);
                         }
                     }
 
                     if (($t = $p->getType()) && !$t->isBuiltin()) {
-                        self::doPreload($t->getName(), $preloaded);
+                        self::doPreload($t->getName(), $preloaded, $loader);
                     }
                 }
 
                 if (($t = $m->getReturnType()) && !$t->isBuiltin()) {
-                    self::doPreload($t->getName(), $preloaded);
+                    self::doPreload($t->getName(), $preloaded, $loader);
                 }
             }
         } catch (\ReflectionException $e) {
             // ignore missing classes
         }
     }
+
+    private static function getClassImports($file)
+    {
+        $tokens = token_get_all(file_get_contents($file));
```

Comment:
I'd rather not do this.
Instead, the list of files should be generated by the PhpDumper.

Reply:
Even doing this, we can only preload about 1/3 of the files.

AFAIK there is no way using reflection to detect imports in a class that is not directly part of the parent, property or methods. E.G

```php
use Some\Example\TestException;

class Foo {
    public function test() {
        throw new TestException();
    }
}
```

Opcache won't preload this file if `Some\Example\TestException` does not exist first, so we need to load it first. And we can't know that this class is used by just using reflection. Which is why I opted to parse the file contents instead, so that we can get all the required dependant classes before the file is parsed by opcache.

Reply:
We don't care to preload this class, it's not on the hot path. We should not do static code analysis to extract symbols, that will end up being the same as loading every single file in vendor. Let the PR focus on what it does: using `opcache_compile_file()`